python/lsst/ts/salkafka/component_producer_set.py

```diff
@@ -180,4 +329,170 @@ def make_argument_parser():
+                self.log.info("Waiting for producers to start")
+                await asyncio.gather(
+                    *[producer.start_task for producer in self.producers]
+                )
+                self.log.info("Running")
+                self.start_task.set_result(None)
+                self.semaphore_file.touch()
+                await self.wait_forever_task
+            except asyncio.CancelledError:
+                self.log.info("Shutting down")
+                for producer in self.producers:
+                    await producer.close()
+
+        if self.semaphore_file.exists():
+            self.semaphore_file.unlink()
+
+        self.log.info("Done")
+
+    async def run_producer_subprocess(
+        self,
+        *,
+        kafka_config,
+        component,
+        topic_names,
+        queue_len=salobj.topics.DEFAULT_QUEUE_LEN,
+    ):
+        """Run a producer for a subset of topics for one SAL component.
+
+        Parameters
+        ----------
+        kafka_config : `KafkaConfig`
+            Kafka configuration.
+        component : `str`
+            Name of a SAL component for which to handle a subset of topics.
+        topic_names : `TopicNames`
+            Topic names in this topic_names.
+        queue_len : `int`, optional
+            Length of the DDS read queue. Must be greater than or equal to
+            `salobj.domain.DDS_READ_QUEUE_LEN`, which is the default.
+        """
+        async with salobj.Domain() as domain, KafkaProducerFactory(
+            config=kafka_config, log=self.log,
+        ) as kafka_factory:
+            self.domain = domain
+            self.kafka_factory = kafka_factory
+
+            self.log.info(
+                f"Creating partial producer for {component}: "
+                f"[add_ackcmd: {topic_names.add_ackcmd}] "
+                f"[commands: {topic_names.commands}] "
+                f"[events: {topic_names.events}] "
+                f"[telemetry: {topic_names.telemetry}] "
+            )
+
+            self.producers = [
+                ComponentProducer(
+                    domain=domain,
+                    component=component,
+                    kafka_factory=kafka_factory,
+                    queue_len=queue_len,
+                    add_ackcmd=topic_names.add_ackcmd,
+                    commands=topic_names.commands,
+                    events=topic_names.events,
+                    telemetry=topic_names.telemetry,
+                )
+            ]
+
+            loop = asyncio.get_running_loop()
+            for s in (signal.SIGTERM, signal.SIGINT, signal.SIGHUP):
+                loop.add_signal_handler(s, self.signal_handler)
+
+            try:
+                self.log.info("Waiting for partial producer to start")
+                await asyncio.gather(
+                    *[producer.start_task for producer in self.producers]
+                )
+                self.log.info("Running")
+                self.done_task.set_result(None)
```

Comment:
Where is `self.done_task` defined? I don't see it anywhere else in the class and since the class is not subclassing any other class, I don't see where this is coming from.

Reply:
Good catch. It must be a holdover. I've removed it.